external-crates/move/crates/move-compiler/src/expansion/aliases.rs

```diff
@@ -149,145 +217,96 @@ impl AliasMapBuilder {
 impl AliasMap {
     pub fn new() -> Self {
         Self {
-            modules: UniqueMap::new(),
-            members: UniqueMap::new(),
-            unused: vec![],
+            aliases: UniqueMap::new(),
+            previous: None,
         }
     }
 
-    pub fn is_empty(&self) -> bool {
-        let Self {
-            modules,
-            members,
-            unused: _,
-        } = self;
-        modules.is_empty() && members.is_empty()
-    }
-
-    fn current_depth(&self) -> usize {
-        self.unused.len()
-    }
-
-    pub fn module_alias_get(&mut self, n: &Name) -> Option<ModuleIdent> {
-        match self.modules.get_mut(n) {
-            None => None,
-            Some((depth_opt, ident)) => {
-                if let Some(depth) = depth_opt {
-                    self.unused[*depth].modules.remove(n);
-                }
-                *depth_opt = None;
-                // We are preserving the name's original location, rather than referring to where
-                // the alias was defined. The name represents JUST the module name, though, so we do
-                // not change location of the address as we don't have this information.
-                // TODO maybe we should also keep the alias reference (or its location)?
-                let sp!(
-                    _,
-                    ModuleIdent_ {
-                        address,
-                        module: ModuleName(sp!(_, module))
-                    }
-                ) = ident;
-                let address = *address;
-                let module = ModuleName(sp(n.loc, *module));
-                Some(sp(n.loc, ModuleIdent_ { address, module }))
+    pub fn get(&mut self, name: &Name) -> Option<AliasEntry> {
+        if let Some((entry, used)) = self.aliases.get_mut(name) {
+            *used = true;
+            // Type parameters are never resolved. We track them to apply appropriate shadowing.
+            if matches!(entry, AliasEntry::TypeParam) {
+                None
+            } else {
+                Some(entry.clone())
             }
+        } else if let Some(previous) = self.previous.as_mut() {
```

Comment:
1) We might want to make this a loop because... Rust 😡 
2) Have you observed a noticeable perf difference with this? Probably not since this is likely... what? like 5 lists deep?
3) How will this work with locals? Will I have to push a new scope on at each `let` binding?

Reply:
> 1. We might want to make this a loop because... Rust 😡

Ah, sure thing.

> 2. Have you observed a noticeable perf difference with this? Probably not since this is likely... what? like 5 lists deep?

First, yeah, it never gets too deep. Second, the previous implementation cloned the whole map every time it pushed a new scope, so I think this will still be a perf win. Third, if it is ever slow we can do path compression: if we ever look something up in a previous, we can update all the intermediary tables to also hold that information, amortizing it to O(1) lookup (at the cost of space).

> 3. How will this work with locals? Will I have to push a new scope on at each let binding?

You will only need to push scopes when entering new blocks, because any lets (or uses) in that block should "fall out" of scope on exit. (Technically, we only need to do this for blocks that have let bindings or use statements in them.) Otherwise, a let should shadow the previous definition, so we just add it to the innermost scope (replacing anything already there).